README.md
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![Imports: isort](https://img.shields.io/badge/%20imports-isort-%231674b1?style=flat&labelColor=ef8336)](https://pycqa.github.io/isort/)
[![Conventional Commits](https://img.shields.io/badge/Conventional%20Commits-1.0.0-%23FE5196?logo=conventionalcommits&logoColor=white)](https://conventionalcommits.org)
[![Semantic Versioning](https://img.shields.io/badge/semver-2.0.0-green)](https://semver.org/spec/v2.0.0.html)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

# AutoAgora Processor

Processes raw `indexer-service` logs from a RabbitMQ queue, saves the results to a PostgreSQL database.

The processing:

- Extracts details about query execution form the log lines: The subgraph deployment, timestamp, query, variables,
execution time and fees paid.
- Normalizes the queries, separates the "query skeleton" from all values and variables.
- Saves the query skeleton (if never seen before), as well as a logs all the information
  about the query execution (including the variable values used).

Refer to [AutoAgora](https://github.com/semiotic-ai/autoagora) for more
details.

## Indexer's guide

### Build

Just build the container!

```sh
docker build -t autoagora-processor .
```

### Usage

Because it is a stateless service, and only consumes from a RabbitMQ queue and adds rows to a PostgreSQL database, the
Autoagora Processor can safely be dynamically scaled horizontally. Note however that it has been observed processing
upwards of 50 queries per second on a single CPU core.

Note that it also needs to connect to the `indexer-agent`'s management API to determine the indexer's current
allocations, and then pull their schemas using the `graph-node` query endpoint.

Configuration:

```txt
usage: autoagora-processor [-h] [--log-level {DEBUG,INFO,WARNING,ERROR,CRITICAL}] --rabbitmq-host
                           RABBITMQ_HOST [--rabbitmq-queue-name RABBITMQ_QUEUE_NAME]
                           [--rabbitmq-queue-limit RABBITMQ_QUEUE_LIMIT]
                           [--rabbitmq-exchange-name RABBITMQ_EXCHANGE_NAME]
                           [--rabbitmq-username RABBITMQ_USERNAME]
                           [--rabbitmq-password RABBITMQ_PASSWORD] --postgres-host POSTGRES_HOST
                           --postgres-database POSTGRES_DATABASE --postgres-username
                           POSTGRES_USERNAME --postgres-password POSTGRES_PASSWORD
                           [--postgres-port POSTGRES_PORT] --graph-node-query-endpoint
                           GRAPH_NODE_QUERY_ENDPOINT --indexer-agent-mgmt-endpoint
                           INDEXER_AGENT_MGMT_ENDPOINT

options:
  -h, --help            show this help message and exit
  --log-level {DEBUG,INFO,WARNING,ERROR,CRITICAL}
                        [env var: LOG_LEVEL] (default: WARNING)
  --rabbitmq-host RABBITMQ_HOST
                        Hostname of the RabbitMQ server used for queuing the GQL logs. [env var:
                        RABBITMQ_HOST] (default: None)
  --rabbitmq-queue-name RABBITMQ_QUEUE_NAME
                        Name of the RabbitMQ queue to pull query-node logs from. [env var:
                        RABBITMQ_QUEUE_NAME] (default: gql_logs_processor)
  --rabbitmq-queue-limit RABBITMQ_QUEUE_LIMIT
                        Size limit of the created RabbitMQ queue. It is discouraged to change that
                        value while the system is running, because it requires manual destruction
                        of the queue and a restart of the whole Auto Agora stack. [env var:
                        RABBITMQ_QUEUE_LIMIT] (default: 1000)
  --rabbitmq-exchange-name RABBITMQ_EXCHANGE_NAME
                        Name of the RabbitMQ exchange query-node logs are pushed to. [env var:
                        RABBITMQ_EXCHANGE_NAME] (default: gql_logs)
  --rabbitmq-username RABBITMQ_USERNAME
                        Username to use for the GQL logs RabbitMQ queue. [env var:
                        RABBITMQ_USERNAME] (default: guest)
  --rabbitmq-password RABBITMQ_PASSWORD
                        Password to use for the GQL logs RabbitMQ queue. [env var:
                        RABBITMQ_PASSWORD] (default: guest)
  --postgres-host POSTGRES_HOST
                        Host of the postgres instance storing the logs. [env var: POSTGRES_HOST]
                        (default: None)
  --postgres-database POSTGRES_DATABASE
                        Name of the logs database. [env var: POSTGRES_DATABASE] (default: None)
  --postgres-username POSTGRES_USERNAME
                        Username for the logs database. [env var: POSTGRES_USERNAME] (default:
                        None)
  --postgres-password POSTGRES_PASSWORD
                        Password for the logs database. [env var: POSTGRES_PASSWORD] (default:
                        None)
  --postgres-port POSTGRES_PORT
                        Port for the logs database. [env var: POSTGRES_PORT] (default: 5432)
  --graph-node-query-endpoint GRAPH_NODE_QUERY_ENDPOINT
                        URL of the indexer's graph-node GraphQL query endpoint. [env var:
                        GRAPH_NODE_QUERY_ENDPOINT] (default: None)
  --indexer-agent-mgmt-endpoint INDEXER_AGENT_MGMT_ENDPOINT
                        URL to the indexer-agent management GraphQL endpoint. [env var:
                        INDEXER_AGENT_MGMT_ENDPOINT] (default: None)
```


CHANGELOG.md
# Changelog

## [1.0.1](https://github.com/semiotic-ai/autoagora-processor/compare/v1.0.0...v1.0.1) (2023-11-06)


### Bug Fixes

* support single-network mode ([beae50e](https://github.com/semiotic-ai/autoagora-processor/commit/beae50ec355fde7888f18e5d0b4dbd630be30405))

## [1.0.0](https://github.com/semiotic-ai/autoagora-processor/compare/v0.4.1...v1.0.0) (2023-10-30)


### ⚠ BREAKING CHANGES

* autoagora-processor needs to communicate with the indexer-agent API (--indexer-agent-mgmt-endpoint) instead of graph-node's PG DB (--graph-postgres-... args are gone).

### Bug Fixes

* Only load schemas for allocated subgraphs ([2a24173](https://github.com/semiotic-ai/autoagora-processor/commit/2a24173696fd46934a686536f0dff82d90fdd626)), closes [#33](https://github.com/semiotic-ai/autoagora-processor/issues/33)

## [0.4.1](https://github.com/semiotic-ai/autoagora-processor/compare/v0.4.0...v0.4.1) (2023-10-11)


### Bug Fixes

* update dependencies ([058847f](https://github.com/semiotic-ai/autoagora-processor/commit/058847f271e5ef2cc1f4e8521d06259a805188a8))


autoagora_processor/main.py
# Copyright 2022-, Semiotic AI, Inc.
# SPDX-License-Identifier: Apache-2.0

import json
import logging
import signal
from datetime import datetime
from hashlib import blake2b
from typing import Any, List, Mapping, Tuple, Union

import pika
from graphql import GraphQLSchema, parse
from graphql.utilities import strip_ignored_characters as gql_strip_ignored_characters
from prometheus_client import Counter, start_http_server
from thegraph_gql_utils.misc import ast2str
from thegraph_gql_utils.tools import (
    build_variable_definitions,
    extract_root_queries,
    factorize,
    prune_query_arguments,
    remove_default_values,
    remove_query_name,
    remove_unknown_arguments,
    remove_values,
    sort,
    substitute_fragments,
)

from autoagora_processor.config import args, init_config
from autoagora_processor.logs_db import LogsDB
from autoagora_processor.query_log_entry import QueryLogEntry
from autoagora_processor.subgraph_schemas import SubgraphSchemas

subgraph_fees_counter = Counter("Fees", "Fees counter for subgraph", ["subgraph"])

init_config()

logs_db = LogsDB(
    dbname=args.postgres_database,
    user=args.postgres_username,
    password=args.postgres_password,
    host=args.postgres_host,
    port=args.postgres_port,
)

subgraph_schemas = SubgraphSchemas()


# Handle SIGTERM for graceful termination
sigterm_received = False


def sigterm_flip(*args):
    global sigterm_received
    logging.info("SIGTERM received! Shutting down...")
    sigterm_received = True


signal.signal(signal.SIGTERM, sigterm_flip)


def normalize_query(
    query: str, variables: Union[str, Mapping[str, Any]], schema: GraphQLSchema
) -> List[Tuple[str, List[str]]]:
    logging.debug("Query before normalization: %s", query)

    if isinstance(variables, (str, bytes)):
        variables_map = json.loads(variables)
    elif isinstance(variables, Mapping):
        variables_map = variables
    else:
        raise RuntimeError

    outputs = []
    query_ast = parse(query)

    query_ast = substitute_fragments(query_ast)
    query_ast = factorize(query_ast)
    query_ast = prune_query_arguments(query_ast)
    query_ast = sort(query_ast)
    query_ast = remove_unknown_arguments(query_ast, schema)
    query_ast, default_values = remove_default_values(query_ast)

    for root_query in extract_root_queries(query_ast):
        # Merge variable values
        variables_merged = {**default_values, **variables_map}

        root_query, removed_variables = remove_values(
            root_query, existing_variables=variables_merged
        )
        root_query = build_variable_definitions(root_query, schema)
        root_query = remove_query_name(root_query)

        outputs += [
            (gql_strip_ignored_characters(ast2str(root_query)), removed_variables)
        ]

    return outputs


def parse_gql_logline(logline: str):
    res = json.loads(logline)

    assert res["msg"] == "Done executing paid query"

    timestamp = datetime.fromtimestamp(res["time"] / 1000)
    query_time_ms: int = res["responseTime"]
    subgraph: str = res["deployment"]
    query_blob = json.loads(res["query"])
    query: str = query_blob["query"]
    query_variables: Mapping[str, Any] = query_blob["variables"]
    fees = int(res["fees"])

    # Increment the prometheus fees counter for the subgraph
    subgraph_fees_counter.labels(subgraph).inc(fees)

    # Normalize query
    schema = subgraph_schemas[subgraph]
    assert schema is not None

    normalized_queries = normalize_query(query, query_variables, schema)
    return [
        QueryLogEntry(
            timestamp=timestamp,
            query_time_ms=query_time_ms if len(normalized_queries) == 1 else None,
            query_variables=removed_variables if removed_variables else None,
            query=query,
            query_hash=blake2b(query.encode(), digest_size=16).digest(),
            subgraph=subgraph,
            fees=fees if len(normalized_queries) == 1 else None,
        )
        for query, removed_variables in normalized_queries
    ]


def callback(channel, _method, _properties, body):
    # Graceful termination
    global sigterm_received
    if sigterm_received:
        channel.stop_consuming()

    try:
        logline = body.decode()
    except:
        logging.exception("Failed to decode log data: %s", body)
        return

    try:
        logging.debug("GQL log line: %s", logline)
        for db_log_entry in parse_gql_logline(logline):
            logs_db.log_query(db_log_entry)
            logging.info("Parsed GQL log entry: %s", db_log_entry)
    except:
        logging.exception("Failed to parse GQL log line: %s", logline)
        return


def main():
    credentials = pika.PlainCredentials(args.rabbitmq_username, args.rabbitmq_password)
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host=args.rabbitmq_host, credentials=credentials)
    )
    channel = connection.channel()

    # (re)-declare the rabbitmq exchange and queue
    channel.exchange_declare(
        exchange=args.rabbitmq_exchange_name,
        exchange_type="fanout",
    )
    channel.queue_declare(
        queue=args.rabbitmq_queue_name,
        arguments={
            "x-max-length": args.rabbitmq_queue_limit,
            "x-overflow": "drop-head",
        },
    )
    channel.queue_bind(
        queue=args.rabbitmq_queue_name, exchange=args.rabbitmq_exchange_name
    )

    channel.basic_consume(
        queue=args.rabbitmq_queue_name,
        on_message_callback=callback,
        auto_ack=True,
    )

    # Start prometheus HTTP server
    start_http_server(8000)

    logging.info("Start consuming from queue...")
    channel.start_consuming()
    logging.info("Terminating...")


if __name__ == "__main__":
    main()


autoagora_processor/logs_db.py
# Copyright 2022-, Semiotic AI, Inc.
# SPDX-License-Identifier: Apache-2.0

import json

import psycopg2 as pg

from autoagora_processor.query_log_entry import QueryLogEntry


class LogsDB:
    def __init__(
        self,
        dbname: str,
        user: str,
        password: str,
        host: str,
        port: str,
    ) -> None:
        self.conn = pg.connect(
            dbname=dbname, user=user, password=password, host=host, port=port
        )
        self.conn.autocommit = True
        self.cur = self.conn.cursor()

        self.cur.execute(
            """
            CREATE TABLE IF NOT EXISTS query_skeletons (
                hash    bytea   PRIMARY KEY,
                query   text    NOT NULL
            )
            """
        )

        self.cur.execute(
            """
            CREATE TABLE IF NOT EXISTS query_logs (
                id              uuid        PRIMARY KEY DEFAULT gen_random_uuid(),
                subgraph        char(46)    NOT NULL,
                query_hash      bytea       REFERENCES query_skeletons(hash),
                timestamp       timestamptz NOT NULL,
                query_time_ms   integer,
                query_variables text
            )
            """
        )

    def log_query(self, query: QueryLogEntry):
        self.cur.execute(
            """
            INSERT INTO query_skeletons (hash, query)
                VALUES (%s, %s)
                ON CONFLICT (hash) DO NOTHING
            """,
            (query.query_hash, query.query),
        )

        self.cur.execute(
            """
            INSERT INTO query_logs (
                    subgraph,
                    query_hash,
                    timestamp,
                    query_time_ms,
                    query_variables
                )
                VALUES (%s, %s, %s, %s, %s)
            """,
            (
                query.subgraph,
                query.query_hash,
                query.timestamp,
                query.query_time_ms,
                json.dumps(query.query_variables),
            ),
        )


autoagora_processor/db_utils.py
# Copyright 2022-, Semiotic AI, Inc.
# SPDX-License-Identifier: Apache-2.0

from typing import Set

import psycopg2
from more_itertools import flatten

from autoagora_processor.config import args


def get_indexed_subgraphs() -> Set[str]:
    conn = psycopg2.connect(
        dbname=args.graph_postgres_database,
        user=args.graph_postgres_username,
        password=args.graph_postgres_password,
        host=args.graph_postgres_host,
        port=args.graph_postgres_port,
    )
    cur = conn.cursor()

    cur.execute(
        """
        select
        d.deployment
        from subgraphs.subgraph_deployment as d,
        subgraphs.subgraph_deployment_assignment as a,
        subgraphs.subgraph_version as v,
        subgraphs.subgraph as g,
        ethereum_networks as n,
        deployment_schemas as s
        where g.id = v.subgraph
        and v.id in (g.pending_version, g.current_version)
        and a.id = s.id
        and s.id = d.id
        and v.deployment = d.deployment
        and n.name = s.network
        and not a.node_id = 'removed';
        """
    )

    return set(flatten(cur.fetchall()))


autoagora_processor/indexer_api.py
# Copyright 2023-, Semiotic AI, Inc.
# SPDX-License-Identifier: Apache-2.0

import logging
from typing import Mapping, Optional, Set

import backoff
from gql import Client, gql
from gql.transport.exceptions import TransportQueryError
from gql.transport.requests import RequestsHTTPTransport
from requests import RequestException

from autoagora_processor.config import args


@backoff.on_exception(backoff.expo, RequestException, max_time=30, logger=logging.root)
def query_indexer_agent(query: str, variables: Optional[Mapping] = None):
    client = Client(
        transport=RequestsHTTPTransport(args.indexer_agent_mgmt_endpoint),
        fetch_schema_from_transport=False,
    )
    result = client.execute(gql(query), variable_values=variables)  # type: ignore
    return result


def get_network_allocated_subgraphs(network: str) -> Set[str]:
    """Get the indexer's subgraph allocations for the given Graph network."""

    result = query_indexer_agent(
        """
        query ($protocolNetwork: String!) {
            indexerAllocations (protocolNetwork: $protocolNetwork) {
                subgraphDeployment
            }
        }
        """,
        variables={
            "protocolNetwork": network,
        },
    )

    return set(e["subgraphDeployment"] for e in result["indexerAllocations"])


def get_allocated_subgraphs() -> Set[str]:
    """
    Get the indexer's subgraph allocations for all Graph networks.

    Will try for both mainnet and arbitrum-one networks. If one of them fails, it
    will ignore it (happens if the indexer-agent is in single network mode).
    If both fail, it will raise the exception.
    """

    networks = ("mainnet", "arbitrum-one")
    results = []

    for network in networks:
        try:
            results += [get_network_allocated_subgraphs(network)]
        except TransportQueryError:
            logging.info(
                f"Failed to get indexer allocations for network '{network}'. Ignoring."
            )
            results += [None]

    if all(r is None for r in results):
        raise RuntimeError(
            f"Failed to query indexer allocations for all Graph networks: {networks}."
        )

    # Replace None's with empty set
    results = [r if r is not None else set() for r in results]

    return set.union(*results)


autoagora_processor/subgraph_schemas.py
# Copyright 2022-, Semiotic AI, Inc.
# SPDX-License-Identifier: Apache-2.0

import logging
from collections.abc import Mapping
from typing import Dict, Iterator, Optional

from gql import Client, gql
from gql.transport.exceptions import TransportQueryError
from gql.transport.requests import RequestsHTTPTransport
from graphql import (
    GraphQLSchema,
    IntrospectionQuery,
    build_client_schema,
    get_introspection_query,
)

from autoagora_processor.config import args
from autoagora_processor.indexer_api import get_allocated_subgraphs


class SubgraphSchemas(Mapping):
    def __init__(self) -> None:
        super().__init__()

        self.schemas: Dict[str, Optional[GraphQLSchema]] = {
            subgraph: self.get_schema_from_graph_node(subgraph)
            for subgraph in get_allocated_subgraphs()
        }
        logging.info("Added schemas for %s subgraphs", len(self.schemas))

    @staticmethod
    def get_schema_from_graph_node(subgraph_ipfs_hash: str) -> Optional[GraphQLSchema]:
        """Get schema string from graph node URL.
        Taken from https://gitlab.semiotic.ai/thegraph/query_generation

        Args:
            subgraph_ipfs_hash (str)

        Returns:
            str: GraphQL schema.
        """

        logging.info("Adding schema for subgraph %s ...", subgraph_ipfs_hash)
        introspection_query = get_introspection_query(descriptions=False)
        client = Client(
            transport=RequestsHTTPTransport(
                f"{args.graph_node_query_endpoint}/subgraphs/id/{subgraph_ipfs_hash}"
            ),
            fetch_schema_from_transport=False,
        )
        try:
            result = IntrospectionQuery(
                client.execute(
                    gql(introspection_query)
                )  # pyright: ignore [reportGeneralTypeIssues]
            )
        except TransportQueryError as e:
            # Should happen when there is an indexing error
            logging.warn(
                "Query error while running introspection query against %s: %s",
                subgraph_ipfs_hash,
                e,
            )
            return None
        # Crash if other error

        return build_client_schema(result)

    def __getitem__(self, subgraph_ipfs_hash: str) -> Optional[GraphQLSchema]:
        if subgraph_ipfs_hash not in self.schemas.keys():
            assert subgraph_ipfs_hash in get_allocated_subgraphs()
            new_schema = self.get_schema_from_graph_node(subgraph_ipfs_hash)
            self.schemas[subgraph_ipfs_hash] = new_schema
            if new_schema:
                logging.info("Added schema for new subgraph %s", subgraph_ipfs_hash)
            else:
                logging.info(
                    "Added None schema for new subgraph %s", subgraph_ipfs_hash
                )
            return new_schema

        return self.schemas[subgraph_ipfs_hash]

    def __iter__(self) -> Iterator[str]:
        return self.schemas.__iter__()

    def __len__(self) -> int:
        return self.schemas.__len__()


autoagora_processor/config.py
# Copyright 2022-, Semiotic AI, Inc.
# SPDX-License-Identifier: Apache-2.0

import logging
from typing import Any

import configargparse


class _Args(configargparse.Namespace):
    def __getattribute__(self, name) -> Any:
        return object.__getattribute__(self, name)


# `args` will contain the global argument values after calling `init_config()`
args = _Args()


def init_config():
    """Parses the arguments from the global argument parser and sets logging config.

    To add arguments, simply add them to the global argument parser from any module, as
    long as the modules are imported before the invocation of that function.

    Argument values are added to the global `args` namespace object declared in this
    module.
    """
    argsparser = configargparse.get_argument_parser()
    #
    # General Arguments
    #
    argsparser.add_argument(
        "--log-level",
        env_var="LOG_LEVEL",
        type=str,
        choices=("DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"),
        default="WARNING",
        required=False,
    )
    #
    # RabbitMQ
    #
    argsparser.add_argument(
        "--rabbitmq-host",
        env_var="RABBITMQ_HOST",
        required=True,
        help="Hostname of the RabbitMQ server used for queuing the GQL logs.",
    )
    argsparser.add_argument(
        "--rabbitmq-queue-name",
        env_var="RABBITMQ_QUEUE_NAME",
        default="gql_logs_processor",
        required=False,
        help="Name of the RabbitMQ queue to pull query-node logs from.",
    )
    argsparser.add_argument(
        "--rabbitmq-queue-limit",
        env_var="RABBITMQ_QUEUE_LIMIT",
        type=int,
        default=1000,
        required=False,
        help="Size limit of the created RabbitMQ queue. It is discouraged to change that "
        "value while the system is running, because it requires manual destruction of the "
        "queue and a restart of the whole Auto Agora stack.",
    )
    argsparser.add_argument(
        "--rabbitmq-exchange-name",
        env_var="RABBITMQ_EXCHANGE_NAME",
        default="gql_logs",
        required=False,
        help="Name of the RabbitMQ exchange query-node logs are pushed to.",
    )
    argsparser.add_argument(
        "--rabbitmq-username",
        env_var="RABBITMQ_USERNAME",
        type=str,
        default="guest",
        required=False,
        help="Username to use for the GQL logs RabbitMQ queue.",
    )
    argsparser.add_argument(
        "--rabbitmq-password",
        env_var="RABBITMQ_PASSWORD",
        type=str,
        default="guest",
        required=False,
        help="Password to use for the GQL logs RabbitMQ queue.",
    )
    #
    # Logs DB
    #
    argsparser.add_argument(
        "--postgres-host",
        env_var="POSTGRES_HOST",
        required=True,
        help="Host of the postgres instance storing the logs.",
    )
    argsparser.add_argument(
        "--postgres-database",
        env_var="POSTGRES_DATABASE",
        required=True,
        help="Name of the logs database.",
    )
    argsparser.add_argument(
        "--postgres-username",
        env_var="POSTGRES_USERNAME",
        required=True,
        help="Username for the logs database.",
    )
    argsparser.add_argument(
        "--postgres-password",
        env_var="POSTGRES_PASSWORD",
        required=True,
        help="Password for the logs database.",
    )
    argsparser.add_argument(
        "--postgres-port",
        env_var="POSTGRES_PORT",
        default=5432,
        required=False,
        help="Port for the logs database.",
    )
    #
    # Graph Node
    #
    argsparser.add_argument(
        "--graph-node-query-endpoint",
        env_var="GRAPH_NODE_QUERY_ENDPOINT",
        required=True,
        help="URL of the indexer's graph-node GraphQL query endpoint.",
    )
    #
    # Indexer Agent API endpoint
    #
    argsparser.add_argument(
        "--indexer-agent-mgmt-endpoint",
        env_var="INDEXER_AGENT_MGMT_ENDPOINT",
        required=True,
        help="URL to the indexer-agent management GraphQL endpoint.",
    )
    argsparser.parse_args(namespace=args)

    logging.basicConfig(format="%(asctime)s - %(name)s - %(levelname)s - %(message)s")
    logging.root.setLevel(args.log_level)
